Refactor Achievement Endpoints for Consistency

NaufalMuh11 opened the pull request:
## Description  
This pull request aims to refactor the achievement endpoints to ensure consistency in HTTP methods and URL parameters. Specifically, the changes include:

- Changing the `PUT` method to `PATCH` for updating achievements.  
- Updating the GET endpoint to use `/:id` instead of `?id=id`.  
- Adding `updateFile` functionality for uploading files to Cloudinary.  
- Fixing related tests to align with the new endpoint structure.  


## Acceptance Criteria  
- The following files should be modified:  
  - `achievementController.js`  
  -  'achivementServices.js'
  - 'achivementValidator.js'
  - 'achivementMiddleware.js'
  - `achievementRoutes.js` 
  - 'utils/uploadFile' 
- Ensure the old code remains intact to support backward compatibility.

---

## Checklist  
- [ ] Refactored `PUT` to `PATCH` for updating achievements.  
- [ ] Changed GET achievement by ID to use `/:id` instead of `?id=id`.  
- [ ] Implemented `updateFile` functionality for Cloudinary uploads.  
- [ ] Updated and fixed test cases to reflect the endpoint changes.  
- [ ] Verified backward compatibility with the old implementation.

---

NaufalMuh11 committed:
refactor: update achievement retrieval to use params instead of query

- Changed the method of finding achievements by ID from query parameters to URL parameters for better RESTful compliance.

NaufalMuh11 committed:
refactor: change update data for achievement service

NaufalMuh11 committed:
feat: create build response for achievement

NaufalMuh11 committed:
feat: create update file function for cloudinary

NaufalMuh11 committed:
refactor: create update achievement for PATCH endpoint

NaufalMuh11 committed:
refactor: update validation and middleware for PATCH achievement endpoint

NaufalMuh11 committed:
refactor: change the PATCH and GET endpoint for testing

NaufalMuh11 committed:
refactor: delete unused upload single

NaufalMuh11 committed:
style: run lint to add semicolon

nugrahaAdhitama commented on src/routes/achievementRoutes.js:
```diff
@@ -26,14 +27,18 @@ router.post(
 
-router.put(
-  '/',
+router.patch(
+  '/:id',
   authentication(),
   checkUpdateFile('logo'),
-  postAchievementValidationRules(),
+  updateAchievementValidationRules(),
   validate,
   achievementController.updateAchievementById,
 );
```
Please make the `router.put` still intact and create a new `router.patch` without deleting the `router.put`

haikalrafifas commented on src/controllers/achievementController.js:
```diff
@@ -58,7 +60,7 @@ exports.getAllAchievements = async (req, res, next) => {
 
 exports.getAchievementById = async (req, res) => {
   try {
-    const { id } = req.query;
+    const { id } = req.params;
```
Please keep the old method of `req.query` intact in the controllers, so that it won't significantly impact Frontend development. This applies to all method inside this controller.

NaufalMuh11 committed:
feat: keep the old method of req query

NaufalMuh11 committed:
refactor: keep the old method of put and get achievement by id using query endpoint

NaufalMuh11 committed:
# refactor: remove default folder value in updateFile function

- Removed the default `folder` value ('logo') in the `updateFile` function to make it more flexible.
- Now the `folder` parameter must be explicitly provided when calling the function.

haikalrafifas committed:
Merge branch 'staging' into 75-refactor-endpoint-achievements
